srml/staking/src/tests.rs

```diff
@@ -503,97 +618,485 @@ fn staking_eras_work() {
+
+		block = 9; // Block 9 => Session 3 => Era 1
+		System::set_block_number(block);
+		Timestamp::set_timestamp(block*5); 
+		Session::check_rotate_session(System::block_number());
+		assert_eq!(Staking::current_era(), 1);
+		assert_eq!(Session::current_index(), 3);
 
-		assert_eq!(Balances::total_balance(&10), 0);
+		// whats left to be shared is the sum of 3 rounds minus the validator's cut.
+		let shared_cut = 3 * session_reward - validator_cut;
+		// Validator's payee is Staked account, 11, reward will be paid here.
+		assert_eq!(Balances::total_balance(&11), validator_initial_balance + shared_cut/2 + validator_cut);
+		// Controller account will not get any reward.
+		assert_eq!(Balances::total_balance(&10), 1);
+		// Rest of the reward will be shared and paid to the nominator in stake.
+		assert_eq!(Balances::total_balance(&2), 20 + shared_cut/2);
 	});
 }
+
+#[test]
+fn staking_ledger_grows_and_shrinks() {
+	// TODO: Show that staking ledger grows with new events
+	// TODO: Show that staking ledger shrinks when user is removed
+}
+
+#[test]
+fn consolidate_unlocked_works() {
+	// TODO: Figure out what it does and then test it
+}
+
+#[test]
+fn bond_extra_works() {
+	// Tests that extra `free_balance` in the stash can be added to stake
+	with_externalities(&mut ExtBuilder::default().build(),
+	|| {
+		// Check that account 10 is a validator
+		assert!(<Validators<Test>>::exists(10));
+		// Check that account 10 is bonded to account 11
+		assert_eq!(Staking::bonded(&11), Some(10));
+		// Check how much is at stake
+		assert_eq!(Staking::ledger(&10), Some(StakingLedger { stash: 11, total: 1000, active: 1000, unlocking: vec![] }));
+
+		// Give account 11 some large free balance greater than total
+		Balances::set_free_balance(&11, 1000000);
+		// Check the balance of the stash account
+		assert_eq!(Balances::free_balance(&11), 1000000);
+
+		// Call the bond_extra function from controller, add only 100
+		assert_ok!(Staking::bond_extra(Origin::signed(10), 100));
+		// There should be 100 more `total` and `active` in the ledger
+		assert_eq!(Staking::ledger(&10), Some(StakingLedger { stash: 11, total: 1000 + 100, active: 1000 + 100, unlocking: vec![] }));
+
+		// Call the bond_extra function with a large number, should handle it
+		assert_ok!(Staking::bond_extra(Origin::signed(10), u64::max_value()));
+		// The full amount of the funds should now be in the total and active
+		assert_eq!(Staking::ledger(&10), Some(StakingLedger { stash: 11, total: 1000000, active: 1000000, unlocking: vec![] }));
+
+	});
+}
+
+#[test]
+fn withdraw_unbonded_works() {
+	// TODO: Learn what it is and test it
+}
+
+#[test]
+fn reporting_misbehaviors_work() {
+	// TODO: Does this code exist?
+}
+
+#[test]
+fn correct_number_of_validators_are_chosen() {
+	// TODO: Check that number is at least minimum, and at most what is set
+	// TODO: Test emergency conditions?
+}
+
+/*
+#[test]
+fn slot_stake_is_least_staked_validator_and_limits_maximum_punishment() {
```

Comment:
@shawntabrizi shouldn't we uncomment this now?

Reply:
Yeah, please commit no commented-out-code.

Reply:
fixed